### DISCUSSION TITLE
Slower loading of MVTLayer in versions 8.5.0 and higher

### DISCUSSION THREAD

**Discussion by DanielaLera** [visgl/deck.gl#6625] [Q&A]
What is the difference for MVTLayer loading mechanism between 8.4.* versions and 8.5.0?
With version 8.4.*, our MVTLayer loads faster than if we upgrade to 8.5.0 or to the latest version. 
This happens for large files that cover large surfaces on the map. When we zoom-in it takes a lot longer until all the details are shown (for the same code, the only difference is the deck.gl version).   

Is it possible to have the same loading speed for MVTLayer in higher versions as in version 8.4.*?  

We want to upgrade from version 8.4.* because we have encountered some rendering issues with multipolygons which contain holes (the holes don't appear as expected) and we've seen this got fixed from version 8.5.0, but we've noticed that the rendering speed in large files is slower than with version 8.4.*.

Thank you!

**Comment by padawannn** [comment]
In the 8.5.0 version, the binary MVTLayer property is false by default and we change it to true in the 8.6.0-beta.1. Maybe, you're using a version where binary is false by default, try to set it to true manually.

[Here](https://deck.gl/docs/api-reference/geo-layers/mvt-layer#binary) you have an explanation about what this property does.

**Reply by DanielaLera** [reply]
Thanks a lot for your answer! We do have the binary property set to true in the MVTLayer, but we don't have the part mentioned in the remarks: "It requires using GeoJsonLayer in the renderSubLayers callback".
We don't have a GeoJsonLayer, we are getting the data as URL templates (pbf files) in the MVTLayer.

**Reply by padawannn** [reply]
@DanielaLera the MVTLayer renders each tile as a GeoJsonLayer by default but there are developers that overwrite the renderSubLayers method to draw each tile using another layer type. If you're not overwriting the renderSubLayers the binary property should work.

**Reply by DanielaLera** [reply]
OK, thank you, I understand. Then I think it's something else, we have the binary property set to true and we didn't overwrite the renderSubLayers method. 
I was looking into the load and loadOptions properties, was wondering what changed between versions 8.4.* and 8.5.0. 
Found this issue related to loaders (MVTLayer and TerrainLayer switch to use worker-only loaders): https://github.com/visgl/deck.gl/pull/5946

I've looked into the info below in quotes and here -> https://deck.gl/docs/developer-guide/loading-data, and tried not to use web workers like here:
import {MVTLoader} from '@loaders.gl/mvt';
new MVTLayer({
  loaders: [MVTLoader],
  loadOptions: {worker: false}
});

 Or to use a custom workedUrl and use the local worked bundle:

new MVTLayer({
  loadOptions: {
    mvt: {
      workerUrl: "node_modules/@loaders.gl/mvt/dist/mvt-loader.worker.js"
    }
  }
}
However, I don't have the mvt-loader.worker.js file under node_modules/@loaders.gl/mvt/dist/, there is just mvt-loader.js, mvt-worker.js and others.

"Note that by default, the `MVTLoader` parses data using web workers. The worker bundle points to the latest published `@loaders.gl/mvt` NPM module on [unpkg.com](unpkg.com). Custom `workerUrl` may be desirable if the application wishes to serve the worker code itself without relying on the CDN. The worker bundle can be located locally in `"node_modules/@loaders.gl/mvt/dist/mvt-loader.worker.js"`. "

**Reply by Pessimistress** [reply]
From 8.4 to 8.5 deck.gl switched to using loaders.gl 3.0.

I'm not following what you are trying to do in the last comment. Can you share your original layer code that has experienced the regression?

**Reply by DanielaLera** [reply]
Please see below the MVTLayer code: 

    return new MVTLayer({
        key: id,
        id: id,
        data: tileURL,
        minZoom: minZoom,
        maxZoom: maxZoom,
        tileSize: 1024,
        getLineColor: (d: MVTLayerProps) => getColor(d.properties[attribute]),
        getFillColor: (d: MVTLayerProps) => getColor(d.properties[attribute]),
        extruded: extruded,
        getElevation: (d: MVTLayerProps) => getElevation(d.properties[attribute] / 1),
        lineWidthMinPixels: 1,
        getPointRadius: 100, //(d) =>d.properties[attribute] / 2,
        opacity: opacity,
        binary: true,
        pickable: false,
        visible: props.visible,
        maxRequests: 48,
        depthTest: false,
        updateTriggers: {
            getLineColor: { colorRange, displayType },
            getFillColor: { colorRange, displayType },
            getElevation: attributeProps
        },
    });

**Reply by Pessimistress** [reply]
I suspect it has to do with your accessors. If you cannot share your dataset, you can try:

- Use `binary: false` and compare 8.4 and 8.5 performance
- Use Deck's `_onMetrics` callback to check if `updateAttributesTime` is taking longer after the dependency bump

**Reply by DanielaLera** [reply]
Thanks for your answer! Unfortunately, I can't share my dataset. I tried with the _onMetrics callback to check the updateAttributesTime. 
I tried the same scenario using the same code, once with deck.gl version 8.4.20 and once with latest version 8.6.8. 
After saving the settings in the application, I zoomed in to see the details in a corner of the layer on the map. With the latest version only after the 280 seconds where updateAttributeTime was 0, I could see the details on the map (first image attached). With version 8.4.20, the details were loaded almost instantly (second image attached). I repeated the tests a few times, and the results are similar.

![LoadLargeDatasetDeckLastVersion](https://user-images.githubusercontent.com/5565940/153421350-4e8ab869-22e2-4451-b946-2d1ac22534e1.JPG)
![LoadLargeDatasetDeckV8 4 20](https://user-images.githubusercontent.com/5565940/153421382-12047c63-157c-4562-b8b5-b3eecf0d1a95.JPG)

**Reply by Pessimistress** [reply]
Have you tried `binary:false`?

**Reply by Pessimistress** [reply]
What is your OS/browser?

**Reply by DanielaLera** [reply]
Windows 10
Chrome Version 98.0.4758.82 (Official Build) (64-bit)

I'm trying now with deck.gl version 8.4.20 and binary:false, but it froze, I can't zoom-in, zoom-out anymore, no errors in console yet. I get page unresponsive.

**Reply by Pessimistress** [reply]
Can you describe what is in your dataset (points, line strings or polygons)?

One change between the two versions is that the binary format now supports polygon with holes. If you have unusually complex polygons that require triangulation that could take the loader a long time (and when you set binary:false this is done on the main thread and cause the app to freeze).

**Reply by DanielaLera** [reply]
The dataset has about 44 multipolygons that overlap, on a surface about 3500km2.

**Reply by Pessimistress** [reply]
It does not matter how big the total area is. You should check the complexity (number of polygon vertices) inside one tile. Each tile is only displayed at its zoom level, so its content is usually simplified to match the resolution that is visible at that particular zoom. Having too much detail inside a single tile does not improve visual quality, just increases loading time. You stated that loading time in v8.6 is longer, but the map is not visually different, which makes me think that your tile includes unnecessary details (especially polygon holes).
